config/cmake/modules/FindHYPRE.cmake

```diff
@@ -28,6 +31,32 @@ if (HYPRE_FOUND)
   return()
 endif()
 
+if (FETCH_TPLS)
+  # define external project and create future include directory so it is present
+  # to pass CMake checks at end of MFEM configuration step
+  set(PREFIX ${CMAKE_BINARY_DIR}/fetch/hypre)
+  include(ExternalProject)
+  ExternalProject_Add(hypre
+    GIT_REPOSITORY https://github.com/hypre-space/hypre.git
+    GIT_TAG cb7597ffd998dc5270c2e32025e799e68048b1cd # Release 2.32.0
+    GIT_SHALLOW TRUE
+    SOURCE_SUBDIR src
+    PREFIX ${PREFIX}
+    CMAKE_CACHE_ARGS
+      -DCMAKE_BUILD_TYPE:STRING=${CMAKE_BUILD_TYPE}
+      -DCMAKE_INSTALL_PREFIX:PATH=${PREFIX})
```

Comment:
We need to translate all MFEM options (CUDA, HIP etc.) to avoid user confusion. I think this will be a quite daunting task.

Reply:
My opinion is that if this new feature doesn't introduce any confusion beyond the `HYPRE_FOUND` workflow, then that's acceptable albeit maybe less than ideal (i.e., no increase in technical debt).  

Does this new feature operate differently than the original workflow if a user specifies those options and then sets `HYPRE_DIR` to an installation of hypre without those options?

To clarify, I agree that having all things consistent is where this new functionality can be taken, but I think that will be better as a series of PRs... what do you think?

Reply:
I'm just thinking about the bug reports where people say
> I configured MFEM with `-DMFEM_USE_CUDA=ON -DFETCH_TPLS=ON` and hypre doesn't work on GPU. Your code is broken.

Reply:
Makes sense.  Some options that come to mind are:
1. I can include a blanket CMake warning here that indicates the hypre installation may not support all the MFEM options (simplest but easiest for user to ignore)
2. I can include CMake errors if `FETCH_TPLS=ON` is used with `MFEM_USE_CUDA=ON`, `MFEM_USE_HIP=ON`, and whatever else we can think of
3. I can add support for the flags we think are most likely to generate those types of messages
4. Some combination of the above

Reply:
My vote would be on option 2 or 3. I think 3 is doable but requires some work.

Reply:
I've added support for CUDA, HIP, and single precision in 89f1dce.  Full disclosure: I didn't test the CUDA and HIP implementations due to my general lack of experience, so hopefully someone else can test them.

I also added a CMake message that spits out what the flags are that hypre will be built with.